File: js/src/structure/structure.test.ts
import { describe, it, expect } from "vitest";
import { loadCase } from "../conform/fixtures.js";
import {
  buildStructure,
  type StructureBlock,
  type StructureIndex,
} from "./structure.js";

// Proven against the shared fixture — every case must match the Python reference
// citenexus.evidence.structure.build_structure exactly. Follows the §4 exemplar:
// load conformance/cases/structure.json, assert deep equality over ALL cases,
// no leniency.
interface StructureCase {
  name: string;
  document_id: string;
  structure_type: string;
  blocks: StructureBlock[];
  expected: StructureIndex;
}

/** Vector counts, pinned. A vector silently dropped from the fixture is a
 *  weakened contract that no per-case assertion can see. */
const EXPECTED_COUNTS: Record<string, number> = { cases: 11 };

describe("structure index conformance", () => {
  const { cases } = loadCase<{ cases: StructureCase[] }>("structure.json");

  it("vector counts are pinned", () => {
    expect({ cases: cases.length }).toEqual(EXPECTED_COUNTS);
  });

  for (const c of cases) {
    it(c.name, () => {
      const doc = {
        document_id: c.document_id,
        structure_type: c.structure_type,
        blocks: c.blocks,
      };
      expect(buildStructure(doc)).toEqual(c.expected);
    });
  }
});

File: golang/structure/structure_test.go
package structure

import (
	"encoding/json"
	"testing"

	"github.com/muthuishere/citenexus/golang/internal/conform"
)

// BuildStructure is proven against the shared fixture — every case must match
// the Python arbiter (citenexus.evidence.structure.build_structure) exactly.
// Load the fixture, assert over ALL cases, no leniency. Produced and expected
// are both marshalled to canonical JSON and compared byte-for-byte.
func TestStructureConformance(t *testing.T) {
	var fixture struct {
		Cases []struct {
			Name          string          `json:"name"`
			DocumentID    string          `json:"document_id"`
			StructureType string          `json:"structure_type"`
			Blocks        []Block         `json:"blocks"`
			Expected      json.RawMessage `json:"expected"`
		} `json:"cases"`
	}
	conform.Case(t, "structure.json", &fixture)

	if len(fixture.Cases) == 0 {
		t.Fatal("no structure cases loaded")
	}

	for _, c := range fixture.Cases {
		var want Index
		if err := json.Unmarshal(c.Expected, &want); err != nil {
			t.Fatalf("%s: unmarshal expected: %v", c.Name, err)
		}
		doc := Doc{
			DocumentID:    c.DocumentID,
			StructureType: c.StructureType,
			Blocks:        c.Blocks,
		}
		got := BuildStructure(doc)

		gotJSON, err := json.Marshal(got)
		if err != nil {
			t.Fatalf("%s: marshal got: %v", c.Name, err)
		}
		wantJSON, err := json.Marshal(want)
		if err != nil {
			t.Fatalf("%s: marshal want: %v", c.Name, err)
		}
		if string(gotJSON) != string(wantJSON) {
			t.Errorf("%s:\n got  %s\n want %s", c.Name, gotJSON, wantJSON)
		}
	}
}
